### PR TITLE
feat: allow templating of the extends field

### PR DESCRIPTION
**Pull request by lstoeferle**
## Changes

Hi everyone,
this PR adds the `extends` field to the allowed field for handlebars templating to add some QOL for self-hosted config presets within the same directory like the self-host-config.

Self-Hosted configuration
```js
module.exports = {
  customEnvVariables: {
    GITLAB_REF: process.env.CI_COMMIT_REF_NAME,
  },
}
```

Repository configuration

```
{
  extends: ["local>foo/bar#{{ env.GITLAB_REF }}"]
}
```

## Context

https://github.com/renovatebot/renovate/discussions/27952

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — CHANGES_REQUESTED** [review]
wrong, doesn't work that way

File: lib/util/template/index.ts
Comment (rarkins) [inline lib/util/template/index.ts:59]: This does not achieve what you want. It exposes extends which is not desired

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by lstoeferle** [comment]
There's one issue with the [replacements:messageFormat-{{package}}-to-@messageformat/{{package}}](https://docs.renovatebot.com/presets-replacements/#replacementsmessageformat-package-to-messageformatpackage) preset, since it contains `{{package}}`. I've added a [presetTemplateIgnoredFields](https://github.com/renovatebot/renovate/pull/27955/files#diff-9e69e6b71dc886fe57d151497b7b070c4a834c781cd288dd128220e3a6e8423dR47) config, which skips the template compilation in case any of the values are present in the template string.

File: lib/util/template/index.ts
Comment (lstoeferle) [reply lib/util/template/index.ts]: Thx, this looked way too easy. Now I got the meaning of that Array.

File: lib/config/presets/index.spec.ts
Comment (viceice) [inline lib/config/presets/index.spec.ts:330]: no

https://github.com/renovatebot/renovate/blob/main/docs/development/best-practices.md#unit-testing

**Comment by lstoeferle** [comment]
> I don't think that templating based on package is a valid concept. extends fields are evaluated during repository intialization, which is before any packages are detected. I think we need to go back to the discussion to understand better what you're trying to do

It's not my intention to do templating base on package 😉 The problem is, that there is an existing preset which is called `replacements:messageFormat-{{package}}-to-@messageformat/{{package}}`, which includes the handlebars `{{package}}`. So in the end there will be `replacements:messageFormat--to-@messageformat/`, since it will be filled with an empty string by default.

I've added a more detailed scenario here: https://github.com/renovatebot/renovate/discussions/27952#discussioncomment-8826555

**Comment by rarkins** [comment]
Let's rename that `messageFormat-{{package}` one - plus add a migration - and ideally add a test that such names don't include handlebars syntax in future

**Comment by lstoeferle** [comment]
> Let's rename that `messageFormat-{{package}` one - plus add a migration - and ideally add a test that such names don't include handlebars syntax in future

Applied the changes. For the test I'm not quite sure how/where to implement that.

File: lib/config/presets/internal/replacements.ts
Comment (viceice) [inline lib/config/presets/internal/replacements.ts:943]:
```suggestion
  title: 'messageFormat-to-scoped',
```
we should use same name format as for other scoped packages.

this needs to be done in a separate PR.

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/config/presets/index.ts
Comment (viceice) [inline lib/config/presets/index.ts:326]: you're passing an empty object so every handlebar template is replaced with an empty string. I don't know how your test can succeed 🤔

File: lib/config/presets/internal/index.spec.ts
Comment (viceice) [inline lib/config/presets/internal/index.spec.ts:48]: there are a lot more international presets to check. those tests should be on a separate PR

File: lib/config/presets/index.ts
Comment (lstoeferle) [reply lib/config/presets/index.ts:326]: `template.compile` uses
```typescript
const data = { ...GlobalConfig.get(), ...input, env };
```
where `input` is the method argument (`{}` in this case). I'm not sure where it makes sense to put custom data in that use case, since most of the time the input comes from `env`.

File: lib/config/presets/internal/index.spec.ts
Comment (lstoeferle) [reply lib/config/presets/internal/index.spec.ts:48]: > there are a lot more international presets to check

This tests iterates over all presets defined in [/lib/config/presets/internal/index.ts#L20](https://github.com/lstoeferle/renovate/blob/f25d6559276f8ad78328a06e35be74d8d2bfd4be/lib/config/presets/internal/index.ts#L20). As far as I can see the object contains ALL internally defined presets.

> those tests should be on a separate PR

Will do :+1:

Comment (lstoeferle) [reply lib/config/presets/internal/index.spec.ts:48]: Moved the test to a [separate PR](https://github.com/renovatebot/renovate/pull/28336)

File: lib/config/presets/index.spec.ts
Comment (viceice) [inline lib/config/presets/index.spec.ts:348]:
```suggestion
      expect(local.getPreset).toHaveBeenCalledOnce();
```

**Review by rarkins — CHANGES_REQUESTED** [review]
Can you add a documentation example of your use case?

**Review by viceice — DISMISSED** [review]

**Comment by lstoeferle** [comment]
I've added it to the docs with an additional usage example.

File: docs/usage/config-presets.md
Comment (rarkins) [inline docs/usage/config-presets.md:258]: It's unclear what is meant by this example. Is this meant to be the example of what not to do? Or a problem?

File: docs/usage/config-presets.md
Comment (rarkins) [inline docs/usage/config-presets.md:271]: Is this your actual use case?

I'm not sure if this will confuse users more than help them. i.e. the 99% who don't need this start wondering if they'll miss something.

I thought your use case was to be able to make the org name relative

Comment (lstoeferle) [reply docs/usage/config-presets.md:271]: Sure I can add a more simple use case

Comment (lstoeferle) [reply docs/usage/config-presets.md:271]: Sth. like this? Using a preset based on organisation:
```json
{
  "extends": [
    "config:recommended",
    "local>{{ env.ORG_SLUG }}/renovate"
  ]
}
```

Comment (rarkins) [reply docs/usage/config-presets.md:271]: your use case is that this used only with env variables?

Comment (lstoeferle) [reply docs/usage/config-presets.md]: We can also use `topLevelOrg` for the use case mentioned above.

File: docs/usage/config-presets.md
Comment (viceice) [inline docs/usage/config-presets.md:247]: are you sure this will work?

File: lib/util/template/index.ts
Comment (viceice) [inline lib/util/template/index.ts:67]: I don't think this should be here

Comment (rarkins) [reply lib/util/template/index.ts:67]: Yes, remove

File: docs/usage/config-presets.md
Comment (rarkins) [reply docs/usage/config-presets.md:247]: I don't think it works currently as only `{}` is passed as the object

**Review by viceice — COMMENTED** [review]
there should be a note that only a small subset of options can be used in template

File: docs/usage/config-presets.md
Comment (HonkingGoose) [inline docs/usage/config-presets.md:237]: We use the word provide twice in the same sentence. Also, after reading, it's not clear what the reader can _do_ with this feature, or why they should care.

I'd like something like this:

You can use x to be flexible with your presets.
This can be handy when you want to do y.
Read link to docs to learn more.

Note how I'm using the word `you` to put the reader into the text.

File: docs/usage/config-presets.md
Comment (HonkingGoose) [inline docs/usage/config-presets.md:242]:
```suggestion
The company is split into multiple organizations:

- `foo`
- `bar`
- `baz`

Each organization has its own opinionated shareable preset.
```

File: docs/usage/config-presets.md
Comment (HonkingGoose) [inline docs/usage/config-presets.md:248]:
```suggestion
    ORGANIZATION_SLUG: process.env.ORGANIZATION_SLUG,
```

File: docs/usage/config-presets.md
Comment (HonkingGoose) [inline docs/usage/config-presets.md:250]:
```suggestion
  extends: ['local>{{ env.ORGANIZATION_SLUG }}/renovate'],
```

File: docs/usage/config-presets.md
Comment (HonkingGoose) [inline docs/usage/config-presets.md:243]:
```suggestion
Renovate is configured to load the preset of each organisation dynamically based on the `ORGANIZATION_SLUG`environment variable.
```

We use American English, which uses a z for words like organization. Please use the z form in the docs and the code.

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/config-presets.md
Comment (lstoeferle) [reply docs/usage/config-presets.md]: Thanks for the valuable feedback 🙌  I've applied your suggestions, please have a look.

File: docs/usage/config-presets.md
Comment (HonkingGoose) [inline docs/usage/config-presets.md:249]:
```suggestion
The company is split into multiple organizations:
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

File: docs/usage/config-presets.md
Comment (rarkins) [inline docs/usage/config-presets.md:255]: Isn't this use case satisfied by https://docs.renovatebot.com/self-hosted-configuration/#inheritconfig ?

Comment (lstoeferle) [reply docs/usage/config-presets.md:255]: Hmm... I didn't know about this. So we probably would need to find another example, since `inheritConfig` does not cover the use-case of adding a dynamic reference ([see PR description](https://github.com/renovatebot/renovate/pull/27955#issue-2188579889)).

Comment (lstoeferle) [reply docs/usage/config-presets.md:255]: @rarkins how about this use case:
Having a central Renovate preset (default.json) located in a repository `renovate/presets`. But actually before blindly pushing changes to main, it would be nice to validate the changes. This can be achieved by using templating:

```javascript
// config.js
module.exports = {
  customEnvVariables: {
    GITLAB_REF: process.env.CI_COMMIT_REF_NAME || "main",
  },
  extends: ["local>renovate/presets#{{ env.GITLAB_REF }}"]
}
```

I found an older discussion, which could benefit from this feature:
https://github.com/renovatebot/renovate/discussions/16108

Comment (lstoeferle) [reply docs/usage/config-presets.md]: Updated the docs, please have another look.

File: docs/usage/config-presets.md
Comment (HonkingGoose) [inline docs/usage/config-presets.md:248]:
```suggestion
The following example shows a self-hosted Renovate preset located in a GitLab repository called `renovate/presets`.
```

**Review by rarkins — DISMISSED** [review]

File: docs/usage/config-presets.md
Comment (HonkingGoose) [inline docs/usage/config-presets.md:256]:
```suggestion
Usually you want to validate the preset before you put it in your Renovate configuration.
```

File: docs/usage/config-presets.md
Comment (HonkingGoose) [inline docs/usage/config-presets.md:257]:
```suggestion
Here is an example of how you can use templating to validate and load the preset on a branch level:
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
I'll let a maintainer check the Renovate code and the configuration examples.

**Review by HonkingGoose — APPROVED** [review]
Sorry for crossing wires on the PR reviews with the lead maintainer. :wink: 

I'll let the maintainer review again. :upside_down_face:

**Comment by lstoeferle** [comment]
@rarkins may I ask you for a final approval? 😇

**Review by rarkins — APPROVED** [review]

**Pull request merged**
